_P177_CM1107.ino

```diff
@@ -0,0 +1,235 @@
+
+
+#define CM1107_ADDR   0x31 // default address
+
+#define  CM1107_CMD_CO2     0x01
+#define  CM1107_CMD_ABC     0x10
+#define  CM1107_CMD_CAL     0x03
+#define  CM1107_CMD_SN      0x1F
+#define  CM1107_CMD_VER     0x1E
+
+#define  CM1107_STATUS_INIT 0x01 //0 1 swapped compared to datasheet
+#define  CM1107_STATUS_OK   0x00
+#define  CM1107_STATUS_ERR  0x02
+#define  CM1107_STATUS_OUT  0x03
+#define  CM1107_STATUS_NCAL 0x05
+
+#define  CM1107_CAL_ON 0x00
+#define  CM1107_CAL_OFF 0x02
+
+
+
+boolean plugin_177_begin()
+{
+  Wire.begin();
+  return(true);
+}
+
+boolean plugin_177_setABC(unsigned char abc)
+  // Set auto calibration period on CM1107
+  // Returns true (1) if successful, false (0) if there was an I2C error
+{
+  // Write auto calibration configuration
+  Wire.beginTransmission(CM1107_ADDR);
+  Wire.write(CM1107_CMD_ABC);
+  Wire.write(100); // defined by data-sheet
+  if (abc)
+  {
+    Wire.write(CM1107_CAL_ON);
+    Wire.write(abc);
+  } else
+  {
+    Wire.write(CM1107_CAL_OFF);
+    Wire.write(7);
+  }
+  // ? make calibration concentration value configurable
+  Wire.write((uint8_t)(400 >> 8));
+  Wire.write((uint8_t)400);
+  Wire.write(100); // defined by data-sheet
+  return (Wire.endTransmission() == 0);
+}
+
+String plugin_177_getABC()
+{
+  I2C_write8(CM1107_ADDR, CM1107_CMD_ABC);
+  //delay(500);
+  uint8_t abc; //replace with function parameter?
+  uint16_t co2;
+  uint8_t checksum;
+
+  Wire.requestFrom(CM1107_ADDR, (byte)8);
+  String log = F("CM1107: ABC: ");
+  checksum = Wire.read(); // expected CM1107_CMD_ABC
+  if (checksum != CM1107_CMD_ABC)
+  {
+    log += F("CMD_ERROR ");
+  }
+  checksum += Wire.read(); // expected 100
+  abc = Wire.read(); // expected 0 or 2
+  checksum += abc;
+  if (abc)
+  { //expected read 1-15
+    abc = Wire.read();
+    checksum += abc;
+    log += F("OFF, period: ");
+    log += abc;
+    abc = 0;
+  } else
+  {
+    abc = Wire.read();
+    checksum += abc;
+    log += F("ON, period: ");
+    log += abc;
+  }
+  log += F(", ppm: ");
+  co2 = Wire.read(); // high byte ppm
+  checksum += co2;
+  co2 = (co2 << 8) | Wire.read(); // low byte ppm
+  checksum += (uint8_t)co2;
+  log += co2;
+  checksum += Wire.read(); // expected 100
+  checksum += Wire.read(); // CS
+  log += F(", checksum: ");
+  if (checksum)
+  {
+    log += F("ERR");
+    log += checksum;
+  } else
+  {
+    log += F("OK");
+  }
+  addLog(LOG_LEVEL_INFO,log);
+  return log;
+}
+
+uint16_t plugin_177_getCO2()
+  // Retrieve CO2 data in ppm
+{
+  I2C_write8(CM1107_ADDR, CM1107_CMD_CO2);
+  uint16_t co2;
+  uint8_t status;
+  uint8_t checksum;
+
+  Wire.requestFrom(CM1107_ADDR, (byte)5);
+  checksum = Wire.read(); // CM1107_CMD_CO2
+  co2 = Wire.read(); // high byte ppm
+  checksum += (uint8_t)co2;
+  co2 = (co2 << 8) | Wire.read(); // low byte ppm
+  checksum += (uint8_t)co2;
+  status = Wire.read(); // status
+  checksum += status;
+  checksum += Wire.read(); //CS
+
+  String log = F("CM1107: CO2 ppm:");
```

Comment:
Please wrap this log creation code in a check for `logLevelActiveFor(LOG_LEVEL_INFO)`
That will prevent unneeded memory allocation.

Reply:
Could you please expand how it's done?

Reply:
```c++
if(logLevelActiveFor(LOG_LEVEL_INFO)) {
  String log = F("....");
  log += bla;
  addLog(LOG_LEVEL_INFO, log);
}
```

Reply:
Thanks. It's now added to the code but compilers does not like it much. I checked other plugins using it and nothing special to include such function shuldn't be needed.
error: 'logLevelActiveFor' was not declared in this scope